### ISSUE TITLE
Add sequence packing in VLM datasets (HF and Preloaded)

### ISSUE THREAD

**Issue by yaoyu-33** [NVIDIA-NeMo/Megatron-Bridge#1106]

**Comment by yaoyu-33** [comment]
## Sequence packing wrapper (design)

### Goal
Create a general-purpose sequence packing wrapper that sits on top of a regular `DataLoader`. It concatenates variable-length sequences into fixed-length packed sequences (with proper `loss_mask`) to reduce padding waste while preserving training semantics. For VLM, it must propagate image-related fields consistently.

### Data flow
dataset -> sampler (index-level ops: DP sharding, shuffling) -> dataloader (collate) -> seq packing wrapper

This keeps sharding/reproducibility decisions inside the sampler, and adds packing as a post-processing step over batches/samples already produced by the original loader.

### Assumptions / required fields
- **text-only**: `input_ids`, `labels`, `loss_mask` (same shape), optional `attention_mask`
- **VLM**: same as above plus image-related fields (e.g., `pixel_values` or pre-extracted `image_embeddings`, `image_sizes`, etc.)
- Sequences can be concatenated; no cross-sample reordering constraints beyond those implied by the sampler.

### Public API (proposed)
```python
class SeqPackingWrapper:
    def __init__(
        self,
        original_loader,
        target_seq_len: int,
        buffer_capacity: int = 4096,
        pack_strategy: str = "greedy",  # or "first_fit_decreasing"
        pad_token_id: int | None = None,
        eos_token_id: int | None = None,
        return_attention_mask: bool = True,
        drop_last: bool = True,
        collate_keep_fields: tuple[str, ...] = (),  # extra fields to pass-through
    ) -> None: ...

    def __iter__(self): ...
    def __len__(self) -> int | None: ...  # optional if derivable
```

### Iteration lifecycle
1) Refill an in-memory buffer with samples from `original_loader` until reaching `buffer_capacity` (or loader exhaustion).
2) Flatten to per-sample items if the original loader emits mini-batches.
3) Apply the packing algorithm to produce packed sequences of length `target_seq_len` (or shorter when necessary). Update `loss_mask` accordingly; optionally create `attention_mask`.
4) Emit packed mini-batches matching the expected downstream batch structure.
5) Repeat refill/pack/emit until the original loader is exhausted.

### Packing algorithm (high level)
- Do not split a single sequence across multiple packed sequences unless explicitly allowed (default: disallow split). This avoids label boundary issues.
- Concatenate sequences back-to-back, optionally inserting `eos_token_id` between sequences if desired.
- `loss_mask` is 1 for positions that should contribute to loss; 0 for padding or disabled tokens (e.g., prompt tokens, separators if needed).
- If a remaining sequence doesn't fit in the current packed sample, start a new one.
- Stop early if adding any sequence would exceed `target_seq_len`; pad to full length if `drop_last=False` or emit shorter sequences if acceptable.

### Reproducibility / sharding
- Shuffling and DP sharding must be handled by the original `sampler` to preserve reproducible order.
- The wrapper must be deterministic given a fixed stream of input samples.

### VLM considerations
- All per-sample image fields must travel together with their paired text sequence(s).
- If multiple text segments share the same image context, keep them within the same packed output when possible; otherwise, disable packing across different image contexts.
- Expose a simple rule: `can_pack_together(sample_a, sample_b)` that returns False when image contexts differ.

### Performance notes
- Keep `buffer_capacity` moderate to limit memory, but large enough to improve packing efficiency; tune empirically.
- Pre-compute lengths from `input_ids` (or `loss_mask.sum`) to speed packing decisions.
- Use vectorized ops where possible when assembling final tensors.

### Pseudocode
```python
class SeqPackingWrapper:
    def __init__(self, original_loader, target_seq_len, buffer_capacity=4096,
                 pack_strategy="greedy", pad_token_id=None, eos_token_id=None,
                 return_attention_mask=True, drop_last=True, collate_keep_fields=()):
        self.original_loader = original_loader
        self.target_seq_len = target_seq_len
        self.buffer_capacity = buffer_capacity
        self.pack_strategy = pack_strategy
        self.pad_token_id = pad_token_id
        self.eos_token_id = eos_token_id
        self.return_attention_mask = return_attention_mask
        self.drop_last = drop_last
        self.collate_keep_fields = set(collate_keep_fields)
        self._buffer = []  # list[dict[str, Tensor]] of single-sample items

    def __iter__(self):
        self._buffer.clear()
        upstream = iter(self.original_loader)
        while True:
            if not self._buffer:
                self._refill_buffer(upstream)
                if not self._buffer:
                    break  # upstream exhausted

            packed_batch = self._pack_from_buffer()
            if packed_batch is None:
                # Not enough samples to form a packed sequence
                if not self.drop_last and self._buffer:
                    packed_batch = self._finalize_partial()
                    yield packed_batch
                break
            yield packed_batch

    def _refill_buffer(self, upstream):
        try:
            while len(self._buffer) < self.buffer_capacity:
                chunk = next(upstream)
                samples = self._to_samples(chunk)  # flatten batch->list[dict]
                self._buffer.extend(samples)
        except StopIteration:
            pass

    def _to_samples(self, item):
        # Convert a batch or a single sample dict to a list of per-sample dicts
        ...

    def _can_pack_together(self, a, b):
        # For VLM, ensure compatible image context; for text-only, usually True
        return True

    def _pack_from_buffer(self):
        # Greedy packing without splitting single samples
        current = self._init_empty()
        i = 0
        while i < len(self._buffer):
            candidate = self._buffer[i]
            if not self._can_pack_together(current, candidate):
                i += 1
                continue

            if self._fits(current, candidate):
                self._append(current, candidate)
                self._buffer.pop(i)
            else:
                i += 1

            if self._is_full(current):
                break

        if self._num_tokens(current) == 0:
            return None
        return self._finalize(current)

    # Helpers: _init_empty, _fits, _append, _is_full, _num_tokens, _finalize, _finalize_partial
```

### Example usage
```python
base_loader = DataLoader(train_dataset, sampler=sampler, collate_fn=collate_fn, batch_size=bs)
packed_loader = SeqPackingWrapper(
    base_loader,
    target_seq_len=4096,
    buffer_capacity=8192,
    pack_strategy="greedy",
    pad_token_id=pad_id,
    eos_token_id=eos_id,
    return_attention_mask=True,
    drop_last=True,
)
for batch in packed_loader:
    train_step(batch)
```

### Open questions / TODOs
- Allow optional cross-sample splitting (for even higher packing) with strict label boundary handling?
- Batch size semantics: emit fixed number of packed sequences per yield vs dynamic?
- Integrate with distributed sampler state across epochs to avoid drift.
